modules/TextureClient.ixx

```diff
@@ -0,0 +1,377 @@
+        return RETURN_MUTEX_LOCK; //waiting 100ms, to wait infinite pass INFINITE
+    }
+    return RETURN_OK;
+}
+
+int TextureClient::UnlockMutex()
+{
+    if (ReleaseMutex(hMutex) == 0) {
+        return RETURN_MUTEX_UNLOCK;
+    }
+    return RETURN_OK;
+}
+
+unsigned long TextureClient::AddFile(TexEntry& entry, const bool compress)
+{
+    if (modded_textures.contains(entry.crc_hash)) {
+        return 0;
+    }
+    auto texture_file_struct = new TextureFileStruct();
+    texture_file_struct->crc_hash = entry.crc_hash;
+    should_update = true;
+    const auto dds_blob = TextureFunction::ConvertToCompressedDDS(entry, compress, dll_path);
+    texture_file_struct->data.assign(static_cast<BYTE*>(dds_blob.GetBufferPointer()), static_cast<BYTE*>(dds_blob.GetBufferPointer()) + dds_blob.GetBufferSize());
+    std::lock_guard lock(mutex);
+    modded_textures.emplace(entry.crc_hash, texture_file_struct);
+    return texture_file_struct->data.size();
+}
+
+unsigned ProcessModfile(TextureClient& client, const std::string& modfile, const bool compress)
+{
+    Message("Initialize: loading file %s... ", modfile.c_str());
+    auto file_loader = ModfileLoader(modfile);
+    auto entries = file_loader.GetContents();
+    if (entries.empty()) {
+        Message("No entries found.\n");
+        return 0;
+    }
+    Message("%zu textures... ", entries.size());
+    unsigned long file_bytes_loaded = 0;
+    for (auto& tpf_entry : entries) {
+        file_bytes_loaded += client.AddFile(tpf_entry, compress);
+    }
+    entries.clear();
+    Message("%d bytes loaded.\n", file_bytes_loaded);
+    return file_bytes_loaded;
+}
+
+void TextureClient::LoadModsFromFile(const char* source)
+{
+    static std::vector<std::string> loaded_modfiles{};
+    static unsigned long loaded_size = 0;
+    Message("Initialize: searching in %s\n", source);
+
+    std::ifstream file(source);
+    if (!file.is_open()) {
+        Warning("LoadModsFromFile: failed to open modlist.txt for reading; aborting!!!");
+        return;
+    }
+    Message("Initialize: found modlist.txt. Reading\n");
+    std::string line;
+    std::vector<std::string> modfiles;
+    while (std::getline(file, line)) {
+        // Remove newline character
+        line.erase(std::ranges::remove(line, '\r').begin(), line.end());
+        line.erase(std::ranges::remove(line, '\n').begin(), line.end());
+
+        if (!std::ranges::contains(loaded_modfiles, line)) {
+            modfiles.push_back(line);
+            loaded_modfiles.push_back(line);
+        }
+    }
+    auto files_size = 0u;
+    for (const auto modfile : modfiles) {
+        if (std::filesystem::exists(modfile)) {
+            files_size += std::filesystem::file_size(modfile);
+        }
+    }
+    std::vector<std::future<unsigned>> futures;
+    for (const auto modfile : modfiles) {
+        futures.emplace_back(std::async(std::launch::deferred, ProcessModfile, std::ref(*this), modfile, files_size > 400'000'000));
```

Comment:
I think 400'000'000 bytes is a sensible value. 450mb load to roughly 850mb of non-compressed memory and around 250mb of compressed memory. With 850mb of texmod data, worst case, we still have 1gb free for gw itself.

Reply:
launching this deferred instead of async because some part of the directxtex library doesn't seem work when using its functions asynchronously